### ISSUE TITLE
use RENAME_EXCHANGE when available, keeping a backup and backupcopy=no

### ISSUE THREAD

**Issue by lilydjwg** [vim/vim#9012]
**Is your feature request about something that is currently impossible or hard to do? Please describe the problem.**

Currently, if 'backup' and backupcopy=auto|no, Vim does:

1. rename filename to the backup file
2. write new content into filename

Between these two steps, there is a time window when filename doesn't exist. This causes trouble when the disk is slow (e.g. a rotational disk) and a (poorly-written) filesystem watcher is setup to compile things as files change, e.g. [the setup for svelte](https://github.com/sveltejs/template) will throw errors when Vim writes a file in the project and never resume.

**Describe the solution you'd like**

Use `renameat2` with `RENAME_EXCHANGE` on Linux, [supported by glibc since version 2.28 (three years ago)](https://en.wikipedia.org/wiki/Glibc#Version_history):

1. write new content into a tempfile
2. exchange filename with tempfile
3. rename tempfile to backup file

**Describe alternatives you've considered**

`set backupcopy=yes`. This will write twice. Also, processes with an open file descriptor to the file may observe file changes, reading both old and new content and breaking in strange ways.

**Additional context**

**Comment by chrisbra** [comment]
sounds interesting. I have made some experimental changes here: https://github.com/chrisbra/vim/tree/gh_9012

It's a bit hard to follow the complex logic in `buf_write()`, so I am not completely sure this is at all correct. But from my test, this works (using debugger and strace).

**Comment by lilydjwg** [comment]
Thanks, but it doesn't work for me.

It creates the temp file at `/tmp/vXXXX/DD`, which is on a different filesystem and causes renameat2 to fail.

**Comment by chrisbra** [comment]
okay, I have made a slight adjustment, to test that the tempfile is on the same device, else it won't use renameat2.

Hm, it should perhaps instead just use the 'backupdirectory'? I'll check this later

**Comment by lilydjwg** [comment]
> it should perhaps instead just use the 'backupdirectory'?

Yes, I think so.

**Comment by k-takata** [comment]
Similarly, the [ReplaceFile](https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-replacefilew) API might be able to be used on Win32.

**Comment by chrisbra** [comment]
I have made it differently, but not sure if this solves your initial problem. It checks if `renameat2` works, then it will write to the backup file and at the end switch the backup file and the original file. But now reading your original issue description, this will still cause problems I suspect. So I think I misunderstood the whole request :(

**Comment by lilydjwg** [comment]
Have you pushed your code? I do not see it on your gh_012 branch.

Writing to the backup file will work on normal cases, but it may leave partially written content as the backup file on a crash. I don't see this a problem myself, but in the past we can assume the backup file is an old but complete version (`nobackupcopy`).

**Comment by chrisbra** [comment]
sorry, forgot to push, should be there now

**Comment by lilydjwg** [comment]
I tried the new code. Unfortunately it doesn't behave as expected: no backup file is kept. It also creates an empty backup file and exchanges it with the original file, leaving a time window when the original file path is empty while writing to the original file's inode (no copy will be available if the system crashes at this point).

The `buf_write` function code is over 2k lines long. Maybe it needs to be splitted smaller to make the code easier to modify.
